File: scripts/filebrowser.mjs
#!/usr/bin/env zx

// @sunbeam.schemaVersion 1
// @sunbeam.title Browse My Computer
// @sunbeam.mode command
// @sunbeam.packageName File Browser

import * as path from "path";
import * as fs from "fs/promises";
import * as os from "os";

$.verbose = false;

const startDir = os.homedir();

const files = await fs.readdir(startDir, { withFileTypes: true });

const items = files.map((file) => ({
  title: file.name,
  subtitle: path.join(startDir, file.name),
  fill: path.join(startDir, file.name),
  actions: [{ type: "open", path: path.join(startDir, file.name) }],
}));

console.log(
  JSON.stringify({
    type: "list",
    list: {
      items,
    },
  })
);

File: examples/file-browser/file-browser.py
#!/usr/bin/env python3

import json
import pathlib
import argparse
import sys

parser = argparse.ArgumentParser()
parser.add_argument("--root", required=True, type=pathlib.Path)
parser.add_argument("--show-hidden", action="store_true")
args = parser.parse_args()

root: pathlib.Path = args.root
entries = root.iterdir()
if not args.show_hidden:
    entries = filter(lambda p: not p.name.startswith("."), entries)

json.dump(
    {
        "type": "list",
        "items": [
            {
                "title": path.name,
                "accessories": [str(root.absolute())],
                "actions": [
                    (
                        {
                            "type": "open-url",
                            "url": f"file://{path.absolute()}",
                            "title": "Open File",
                        }
                        if path.is_file()
                        else {
                            "type": "run-command",
                            "command": "browse-files",
                            "title": "Browse Directory",
                            "with": {"root": str(path.absolute())},
                        }
                    ),
                    {
                        "type": "copy-text",
                        "title": "Copy Path",
                        "shorcut": "ctrl+y",
                        "text": str(path.absolute()),
                    },
                    {
                        "type": "run-command",
                        "title": "Delete File",
                        "shortcut": "ctrl+d",
                        "command": "delete-file",
                        "with": {"path": str(path.absolute())},
                    },
                ],
            }
            for path in sorted(entries, key=lambda p: p.name)
        ],
    },
    sys.stdout,
)
